here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/AbstractTask.java

```diff
@@ -398,63 +402,61 @@ protected final void unlock() {
-    final double authorThreshold = (double) percentageThresholdPerPrinciple / 100;
+    final long LIMIT = AbstractTask.limit.get();
     lockAndRequireNew();
     try {
-      do {
-        final long threadCount = AbstractTask.threadCount.get();
-        assert threadCount >= 0L;
-        String author = context.getAuthor();
-        if (author != null) {
-          AtomicInteger authorUsage = AbstractTask.authorUsageMap.getOrDefault(author, new AtomicInteger(0));
-          int authorLimit = (int) (LIMIT * authorThreshold);
-          if (!internal && authorUsage.get() >= authorLimit) {
-            log.info(
-                "NAKSHA_ERR_REQ_LIMIT_4_PRINCIPAL - [Request Limit breached for Principal => appId,author,limit,crtValue] - ReqLimitForPrincipal {} {} {} {}",
-                context.getAppId(),
-                author,
-                authorLimit,
-                authorUsage);
-            throw new TooManyTasks(0, (long) (authorLimit), false);
-          }
-        }
-        if (!internal && threadCount >= LIMIT) {
+      long myIncrementedValue = AbstractTask.threadCount.updateAndGet(l -> {
+        if (!internal && AbstractTask.threadCount.get() >= LIMIT) {
           log.info(
               "NAKSHA_ERR_REQ_LIMIT_4_INSTANCE - [Request Limit breached for Instance => appId,author,limit,crtValue] - ReqLimitForInstance {} {} {} {}",
               context.getAppId(),
-              author,
+              getAuthorName(),
               LIMIT,
               threadCount);
-          throw new TooManyTasks(LIMIT, 0, true);
+          String errorMessage = "Maximum number of concurrent tasks reached for instance (" + LIMIT + ")";
+          throw new TooManyTasks(errorMessage);
         }
-        if (AbstractTask.threadCount.compareAndSet(threadCount, threadCount + 1)) {
-          if (author != null) {
-            AbstractTask.incrementAuthorUsage(author);
-          }
-          try {
-            state.set(State.START);
-            final Future<RESULT> future = threadPool.submit(this::init_and_execute);
-            return future;
-          } catch (RejectedExecutionException e) {
-            throw new TooManyTasks();
-          } catch (Throwable t) {
-            AbstractTask.threadCount.decrementAndGet();
-            if (author != null) {
-              AbstractTask.decrementAuthorUsage(author);
-            }
-            log.atError()
-                .setMessage("Unexpected exception while trying to fork a new thread")
-                .setCause(t)
-                .log();
-            throw new RuntimeException("Internal error while forking new worker thread", t);
-          }
-        }
-        // Conflict, two threads concurrently try to fork.
-      } while (true);
+        return l + 1;
+      });
+      assert myIncrementedValue >= 0L;
+      String principal = getPrincipalName();
+      if (principal != null) {
+        AtomicInteger authorUsageValue = principalUsageMap.merge(
```

Comment:
There is benefit of having optimistic locking (old logic) over pessimistic one (new logic) in case of concurrency. We discussed last time that we should move to pessimistic one only if we find optimistic one too complex to manage. I would like to understand what part was complicated here before we choose new method.
It should be OK, if principal limit is pessimistic one, but the global limit with pessimistic locking can add locking bottleneck.

Reply:
Do not use merge and especially, you added side effects into merge. Please use the following strategy (pseudo code):

```
static val principalUsage = ConcurrentHashMap<String,Int>();
...
val principalId = context.getPrincipalId();
val limit = principalManager.getLimit(context);
if (limit <= 0) throw "principal blocked"
while (true) {
  val counter = principalUsage.get( principalId );
  if (counter == null) {
    // first
    counter =  1;
    var existing = principalUsage.putIfAbsent(principalId, counter);
    if (existing != null) continue; // Repeat, conflict with other thread
    break; // We have our slot
  }
  // increment counter
  if (counter+1 > limit) throw "too many requests"
  if (principalUsage.replace(principalId, counter, counter+1)) break;
  // failed, conflict, repeat
}
// now, enter a try block to ensure that whatever happens, eventually we release our allocated principal slot
try {
  // reserve worker from pool and do what was there before, no code change needed in existing code!
} finally {
   // release principal slot
  while (true) {
    val current = principalUsage.get( principalId );
    if (current == null or current <= 1) error: must not happen
    if (current == 1) {
      // last man standing, remove from map
      // if this fails, conflict, repeat, because another thread tried!
      if (!principalUsage.remove(principalId, current)) continue;
      // done, released
      break;
    } 
    // decrement usage
    if (principalUsage.replace(principalId, current, current-1)) break;
    // failed, repeat, conflict with other thread
  }
}
```
This as well automatically cleans the map, so we do not collect garbage.

Reply:
Fixed.